src/renderer/atlas/AtlasEngine.cpp

```diff
@@ -0,0 +1,1449 @@
+        _sr.textAnalyzer = textAnalyzer.query<IDWriteTextAnalyzer1>();
+    }
+
+    _sr.isWindows10OrGreater = IsWindows10OrGreater();
+
+#ifndef NDEBUG
+    // If you run the Host.EXE project inside Visual Studio it'll have a working directory of
+    // $(SolutionDir)\src\host\exe. This relative path will thus end up in this source directory.
+    _sr.sourceCodeWatcher = wil::make_folder_change_reader_nothrow(SHADER_SOURCE_DIRECTORY, false, wil::FolderChangeEvents::FileName | wil::FolderChangeEvents::LastWriteTime, [this](wil::FolderChangeEvent, PCWSTR path) {
+        if (til::ends_with(path, L".hlsl"))
+        {
+            auto expected = INT64_MAX;
+            const auto invalidationTime = std::chrono::steady_clock::now() + std::chrono::milliseconds(100);
+            _sr.sourceCodeInvalidationTime.compare_exchange_strong(expected, invalidationTime.time_since_epoch().count(), std::memory_order_relaxed);
+        }
+    });
+#endif
+}
+
+#pragma region IRenderEngine
+
+// StartPaint() is called while the console buffer lock is being held.
+// --> Put as little in here as possible.
+[[nodiscard]] HRESULT AtlasEngine::StartPaint() noexcept
+try
+{
+    if (_api.hwnd)
+    {
+        RECT rect;
+        LOG_IF_WIN32_BOOL_FALSE(GetClientRect(_api.hwnd, &rect));
+        std::ignore = SetWindowSize({ rect.right - rect.left, rect.bottom - rect.top });
+
+        if (WI_IsFlagSet(_api.invalidations, ApiInvalidations::Title))
+        {
+            LOG_IF_WIN32_BOOL_FALSE(PostMessageW(_api.hwnd, CM_UPDATE_TITLE, 0, 0));
+            WI_ClearFlag(_api.invalidations, ApiInvalidations::Title);
+        }
+    }
+
+    // It's important that we invalidate here instead of in Present() with the rest.
+    // Other functions, those called before Present(), might depend on _r fields.
+    // But most of the time _invalidations will be ::none, making this very cheap.
+    if (_api.invalidations != ApiInvalidations::None)
+    {
+        RETURN_HR_IF(E_UNEXPECTED, _api.cellCount == u16x2{});
+
+        if (WI_IsFlagSet(_api.invalidations, ApiInvalidations::Device))
+        {
+            _createResources();
+        }
+        if (WI_IsFlagSet(_api.invalidations, ApiInvalidations::SwapChain))
+        {
+            _createSwapChain();
+        }
+        if (WI_IsFlagSet(_api.invalidations, ApiInvalidations::Size))
+        {
+            _recreateSizeDependentResources();
+        }
+        if (WI_IsFlagSet(_api.invalidations, ApiInvalidations::Font))
+        {
+            _recreateFontDependentResources();
+        }
+        if (WI_IsFlagSet(_api.invalidations, ApiInvalidations::Settings))
+        {
+            _r.selectionColor = _api.selectionColor;
+            WI_SetFlag(_r.invalidations, RenderInvalidations::ConstBuffer);
+        }
+
+        // Equivalent to InvalidateAll().
+        _api.invalidatedRows = invalidatedRowsAll;
+    }
+
+#ifndef NDEBUG

```

Comment:
ifndef NDEBUG feels like a brain teaser vs ifdef DEBUG

Reply:
Right, but `_DEBUG` is a non-standard macro, whereas `NDEBUG` is official C++ to say "assertions are disabled" (which is usually the same as "release build").

Reply:
NNNNN Okay fine.